credential_pop typ with status-attestation-request+cwt + huge editorials (#37)
peppelinux/draft-demarco-oauth-status-attestations: draft-demarco-oauth-status-attestations.md

# Introduction

> Status Assertions ensure the non-revocation of digital

Reviewer: is there interest in suspension, revocation or other dynamic state? see https://w3c.github.io/vc-bitstring-status-list/#example-example-statuslistcredential-using-more-complex-entries

Some parties have use cases that include dynamic state that is an enumeration instead of a boolean.

Reviewer 2: thank you, moved to an issue

Reviewer 2: https://github.com/peppelinux/draft-demarco-oauth-status-attestations/issues/49

This specification uses the terms "End-User", "Entity" as defined by
OpenID Connect Core [@OpenID.Core], the term "JSON Web Token (JWT)"

> defined by JSON Web Token (JWT) {{RFC7519}}, the term "CBOR Web Token (CWT)" defined by CBOR Web Token (CWT) {{RFC8392}}.

Reviewer 2: ```suggestion
defined by JSON Web Token (JWT) {{RFC7519}}, the term "CBOR Web Token (CWT)" defined in {{RFC8392}}.
```

- The request MUST contain the base64url encoded hash value of the Digital Credential,
for which the Status Assertion is requested, and enveloped in a signed
Status Assertion Request object.
- The Status Assertion Request object MUST be signed with the private key corresponding
to the confirmation claim assigned by the Issuer and contained within
the Digital Credential.

When the JWT or CWT format are used, the JWT/CWT MUST contain the parameters defined in the following table.

| Header | Description | Reference |
| --- | --- | --- |
| **typ** | It MUST be set to `status-attestation+jwt` when JWT format is used. It MUST be set to `status-attestation+cwt` when CWT format is used. | {{RFC7516}} Section 4.1.1 |
| **alg** | A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST NOT be set to `none` or any symmetric algorithm (MAC) identifier. | {{RFC7516}} Section 4.1.1 |
| **kid** | Unique identifier of the `JWK or` `Cose_Key` used for the signature of the Status Attestation Request, it MUST match the one contained in the Credential. | {{RFC7515}} |

| Payload | Description | Reference |
| --- | --- | --- |

> | **iss** | Wallet identifier. The Wallet identifier value is supposed to be used for identifying a wallet; therefore, it is out of scope for this specs, considering that it may simply be an ephemeral value. | {{RFC9126}}, {{RFC7519}} |

Reviewer 2: ```suggestion
| **iss** | Status Assertion Request Issuer identifier. The value is supposed to be used for identifying the Wallet that has issued the request. It is out of scope for this document defining how this value must be set. | {{RFC9126}}, {{RFC7519}} |
```

> | **aud** | It MUST be set with the Credential Issuer Status Attestation endpoint URL as value that identify the intended audience | {{RFC9126}}, {{RFC7519}} |
> | **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be superior to the value set for `iat`  . | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |

Reviewer 2: ```suggestion
| **aud** | It MUST be set with the Credential Issuer Status Attestation endpoint URL as value that identify the intended audience. | {{RFC9126}}, {{RFC7519}} |
| **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be superior to the value set for `iat` . | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |
```

| **iat** | UNIX Timestamp with the time of JWT/CWT issuance. | {{RFC9126}}, {{RFC7519}} |
| **jti** | Unique identifier for the JWT.  | {{RFC7519}} Section 4.1.7 |
| **credential_hash** | Hash value of the Digital Credential the Status Attestation is bound to. | this specification |
| **credential_hash_alg** |  The Algorithm used of hashing the Digital Credential to which the Status Attestation is bound. The value SHOULD be set to `sha-256`. | this specification |

Below is a non-normative example of a Status Assertion Request with
the JWT headers and payload are represented without applying signature and
encoding, for better readability:

~~~
{
    "alg": "ES256",
    "typ": "status-assertion-request+jwt",
    "kid": $CREDENTIAL-CNF-JWKID
}
.
{
    "iss": "0b434530-e151-4c40-98b7-74c75a5ef760",
    "aud": "https://issuer.example.org/status-assertion-endpoint",
    "iat": 1698744039,
    "exp": 1698830439,
    "jti": "6f204f7e-e453-4dfd-814e-9d155319408c",
    "credential_hash": $Issuer-Signed-JWT-Hash

>     "credential_hash_alg": "sha-256",

Reviewer 2: ```suggestion
    "credential_hash_alg": "sha-256"
```

the member `status_assertion_requests`.

The `status_assertion_requests` MUST be set with an array of strings, where
each string within the array represents a Digital Credential
Status Assertion Request object.

The Credential Issuer that receives the Status Assertion Request object
MUST validate that the Wallet Instance making the request is
authorized to request Status Assertions.
Therefore the following requirements MUST be satisfied:

- The Credential Issuer MUST verify the compliance of all elements in the `status_assertion_requests` object
using the confirmation method contained within the Digital Credential where the Status Assertion Request
object is referred to;
- The Credential Issuer MUST verify that it is the legitimate Issuer of the Digital Credential
to which each Status Assertion Request object refers.

# Status Assertion Response

The response MUST include a JSON object with a member
named `status_assertion_responses`, which contains the
Status Assertions and or the Status Assertion Errors
related to the request made by the
Wallet Instance. In the non-normative example below is
represented an HTTP Response with the
`status_assertion_responses` JSON member:

~~~

> HTTP/1.1 201 Created

Reviewer 2: ```suggestion
HTTP/1.1 200 Created
```

Content-Type: application/json

{
    "status_assertion_responses": ["${base64url(json({typ: status-assertion+jwt, ...}))}.payload.signature", ... ]
}
~~~

The member `status_assertion_responses` MUST be an array of strings,
where each of them represent a Status Assertion Response object,
as defined in
[the section Status Assertion](#status-assertion) or a Status Assertion Error object,
as defined in [the section Status Error](#status-assertion-error).

For each entry in the `status_assertion_responses` array, the following requirements are met:
- Each element in the array MUST match the corresponding element in the request array at
the same position index to which it is related, eg: "[requestAboutA, requestAboutB]" produces "[responseAboutA, responseErrorAboutB]".
- Each element MUST contain the error or the status of the assertion using the `typ` member.
set to "status-assertion+{jwt,cwt}" or "status-assertion-error+{jwt,cwt}", depending by the object type.

> - The corresponding entry in the response MUST be of the same type as requested. For example,

Reviewer 2: ```suggestion
- The corresponding entry in the response MUST be of the same data format as requested. For example,
```

> | JOSE Header | Description | Reference |

Reviewer 2: ```suggestion
| Header Parameter Name | Description | Reference |
```

| --- | --- | --- |

> | **typ** | It MUST be set to `status-attestation-request+jwt` or `status-attestation-request+cwt`. | {{RFC7516}} Section 4.1.1 |

Reviewer 2: we should say here in which cases, when JWT is must be set to ... when CWT ...

Reviewer 2: if this is a JOSE header it cannot be CWT

this is related to JOSE

| **alg** | A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST NOT be set to `none` or any symmetric algorithm (MAC) identifier. | {{RFC7516}} Section 4.1.1 |
| **kid** | Unique identifier of the JWK used for the signature of the Status Attestation Request. It MUST match the one contained in the Credential `cnf.jwk`. | {{RFC7515}} |

> | JOSE Payload | Description | Reference |

Reviewer 2: ```suggestion
| Payload Parameter Name | Description | Reference |
```

~~~

> As exemplified for the JWT format, below is a non-normative example of a Credential proof of possession in CWT format. The CWT headers and payload are presented without applying signature and encoding for improved readability:

Reviewer 2: please mention that the prosed non normative example is provided in an unencoded diagnostic representation ...

Author: The commit that aims to resolve this is: https://github.com/peppelinux/draft-demarco-oauth-status-attestations/pull/37/commits/5980f052b048e7d7cc6aad338e7f6386646e00e5

MUST demonstrate the proof of possession of the Digital Credential to
the Credential Issuer.

The proof of possession is made by enclosing the Digital Credential

> in an object (JWT or CWT) using the same confirmation method referenced in RFC7800.

Reviewer 2: the digital credential is not enclosed, an hash of it is provided in the status assertion request and the same hash in the status assertion response

| Payload | Description | Reference |
| --- | --- | --- |
| **iss** | Wallet identifier. The Wallet identifier is out of the scope of this specs considering that it may be just a
framework that renders the Wallet  identifier itself ephemeral. | {{RFC9126}}, {{RFC7519}} |
| **aud** | It MUST be set with the Credential Issuer Status Attestation endpoint URL as value that identify the intended audience | {{RFC9126}}, {{RFC7519}} |
| **exp** | UNIX Timestamp with the expiration time of the JWT. It MUST be superior to the value set for `iat`  . | {{RFC9126}}, {{RFC7519}}, {{RFC7515}} |
| **iat** | UNIX Timestamp with the time of JWT/CWT issuance. | {{RFC9126}}, {{RFC7519}} |
| **jti** | Unique identifier for the JWT.  | {{RFC7519}} Section 4.1.7 |
| **credential_hash** | Hash value of the Digital Credential the Status Attestation is bound to. | this specification |
| **credential_hash_alg** |  The Algorithm used of hashing the Digital Credential to which the Status Attestation is bound. The value SHOULD be set to `sha-256`. | this specification |

> Below is a non-normative example of a Credential proof of possession with

Reviewer 2: ```suggestion
Below is a non-normative example of a Status Assertion Request with
```

the Credential Issuer.

The proof of possession is made by enclosing the Digital Credential in an

> object (JWT or CWT) signed with the private key that its related public key is referenced in the Digital Credential.

Reviewer 2: I would rephrase saying using the same confirmation method referenced ... 

therefore using the elements defined in RFC7800

Author: The commit https://github.com/peppelinux/draft-demarco-oauth-status-attestations/pull/37/commits/5f23e8e156d732e1d2fbad89b81b62b7952f69fd aims to resolve this comment.

When the JWT or CWT format are used, the JWT/CWT MUST contain the parameters defined in the following table.

| Header | Description | Reference |
| --- | --- | --- |
| **typ** | It MUST be set to `status-attestation+jwt` when JWT format is used. It MUST be set to `status-attestation+cwt` when CWT format is used. | {{RFC7516}} Section 4.1.1 |
| **alg** | A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST NOT be set to `none` or any symmetric algorithm (MAC) identifier. | {{RFC7516}} Section 4.1.1 |

> | **kid** | Unique identifier of the JWK or CWT used for the signature of the Status Attestation Request, it MUST match the one contained in the Credential. | {{RFC7515}} |

Reviewer 2: ```suggestion
| **kid** | Unique identifier of the JWK or "Cose Key" used for the signature of the Status Attestation Request, it MUST match the one contained in the Credential. | {{RFC7515}} |
```

JWT != JWK and therefore != CWT

please check Cose key if good.

| Payload | Description | Reference |
| --- | --- | --- |

> | **iss** | Wallet identifier. | {{RFC9126}}, {{RFC7519}} |

Reviewer 2: ```suggestion
| **iss** | Wallet identifier. | {{RFC9126}}, {{RFC7519}} |
```

concerns here, since a wallet identifier can be a vector of traceability. We need further explaination about the evidence that this identifier is not unique and absolute, preventing the traceability of the wallet.

> | JOSE Header | Description | Reference |

Reviewer 2: ```suggestion
| Header Parameter Name | Description | Reference |
```

| --- | --- | --- |
| **alg** | A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST NOT be set to `none` or to a symmetric algorithm (MAC) identifier. | {{RFC7515}}, {{RFC7517}} |
| **typ** | It MUST be set to `status-attestation+jwt` or `status-attestation-request+cwt`. | {{RFC7515}}, {{RFC7517}} and this specification |
| **kid** | Unique identifier of the Issuer JWK. | {{RFC7515}} |

> | JOSE Payload | Description | Reference |

Reviewer 2: ```suggestion
| Payload Parameter Name | Description | Reference |
```

| JOSE Header | Description | Reference |
| --- | --- | --- |
| **alg** | A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST NOT be set to `none` or to a symmetric algorithm (MAC) identifier. | {{RFC7515}}, {{RFC7517}} |

> | **typ** | It MUST be set to `status-attestation+jwt` or `status-attestation-request+cwt`. | {{RFC7515}}, {{RFC7517}} and this specification |

Reviewer 2: as mentioned before

| Payload | Description | Reference |
| --- | --- | --- |

> | **iss** | Wallet identifier. The Wallet identifier is out of the scope of this specs considering that it may be just a

Reviewer 2: the wallet identifier is not a framework

Reviewer 3: The last sentence ("considering that....") is not clear to me. Is there a reason why we decided to emphasize that the Wallet identifier is out of the scope of this spec? I suggest removing this sentence., if you agree.
@OR13 @marinaado

Reviewer 2: @fmarino-ipzs the way a wallet can be identified or the value to be supposed to be used for identify a wallet is definitively out of the scope of this spec

| **exp** | UNIX Timestamp with the expiry time of the Status Assertion. | {{RFC9126}}, {{RFC7519}} |
| **credential_hash** | Hash value of the Digital Credential the Status Assertion is bound to. | this specification |
| **credential_hash_alg** | The Algorithm used of hashing the Digital Credential to which the Status Assertion is bound. The value SHOULD be set to `sha-256`. | this specification |

> | **cnf** | JSON object containing the cryptographic key binding. It MUST match with the one provided within the related Digital Credential. | {{RFC7800}} Section 3.1 |

Reviewer 2: it is not the cnf itself bu the sub-member contained in it, such as jwk for JWT and Cose Key for CWT, while there may be other parameters like `did` for instance

this must be included in this description

Author: The commit that aims to resolve this is: https://github.com/peppelinux/draft-demarco-oauth-status-attestations/pull/37/commits/5980f052b048e7d7cc6aad338e7f6386646e00e5

}
~~~

> As exemplified for the JWT format, below is a non-normative example of a Credential proof of possession in CWT format provided in an unencoded diagnostic representation. The CWT headers and payload are presented without applying signature and encoding for improved readability:

Reviewer 2: ```suggestion
As exemplified for the JWT format, below is a non-normative example of a Status Assertion Request object in CWT format provided in CBOR diagnostic notation format, as defined in {{rfc8152}} where the CWT headers and payload are represented without applying signature and encoding for better readability:
```

~~~
   {
    / protected / << {
      / alg / 1: -7 / ES256 /
      / typ / 16: -7 / status-attestation-request+cwt /
      / kid / 4: h'3132' / $ISSUER-CWKID /
    } >>,
    / unprotected / {
    },
    / payload / << {
      / iss    / 1: 0b434530-e151-4c40-98b7-74c75a5ef760 /,
      / iat    / 6: 1698744039 /,
      / exp    / 4: 1698830439 /,
      / credential_hash 8: $Issuer-Signed-JWT-Hash /,
      / credential_hash_alg 9: sha-256 /,
      / cnf:{
          "COSE_Key":{...}
      }
    } >>,
   }
~~~

The Status Assertion MUST contain the following claims when the JWT format is used.

| Header Parameter Name | Description | Reference |
| --- | --- | --- |
| **alg** | A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. It MUST NOT be set to `none` or to a symmetric algorithm (MAC) identifier. | {{RFC7515}}, {{RFC7517}} |
| **typ** | It MUST be set to `status-attestation+jwt` when JWT format is used. It MUST be set to `status-attestation+cwt` when CWT format is used. | {{RFC7515}}, {{RFC7517}} and this specification |

> | **kid** | Unique identifier of the Issuer JWK or CWT. | {{RFC7515}} |

Reviewer 2: JWT != JWK therefore != CWT

| Payload Parameter Name | Description | Reference |
| --- | --- | --- |
| **iss** | It MUST be set to the identifier of the Issuer. | {{RFC9126}}, {{RFC7519}} |
| **iat** | UNIX Timestamp with the time of the Status Assertion issuance. | {{RFC9126}}, {{RFC7519}} |
| **exp** | UNIX Timestamp with the expiry time of the Status Assertion. | {{RFC9126}}, {{RFC7519}} |
| **credential_hash** | Hash value of the Digital Credential the Status Assertion is bound to. | this specification |
| **credential_hash_alg** | The Algorithm used of hashing the Digital Credential to which the Status Assertion is bound. The value SHOULD be set to `sha-256`. | this specification |

> | **cnf** | JSON object containing the cryptographic key binding. The sub-member contained within `cnf` member, such as `jwk` for JWT and `Cose_Key` for CWT, MUST match with the one provided within the related Digital Credential. Additionally, there may be other parameters present, such as `did`, for instance. | {{RFC7800}} Section 3.1 |

Reviewer 2: ```suggestion
| **cnf** | JSON object containing the cryptographic key binding. The sub-member contained within `cnf` member, such as `jwk` for JWT and `Cose_Key` for CWT, MUST match with the one provided within the related Digital Credential. Other confirmation methods can be utilized when the referenced Digital Credential supports them, in accordance with the relevant standards. | {{RFC7800}} Section 3.1 |
```

| **exp** | UNIX Timestamp with the expiry time of the Status Assertion. | {{RFC9126}}, {{RFC7519}} |
| **credential_hash** | Hash value of the Digital Credential the Status Assertion is bound to. | this specification |
| **credential_hash_alg** | The Algorithm used of hashing the Digital Credential to which the Status Assertion is bound. The value SHOULD be set to `sha-256`. | this specification |

> | **cnf** | JSON object containing the cryptographic key binding. The sub-member contained within `cnf` member, such as `jwk` for JWT and `Cose_Key` for CWT, MUST match with the one provided within the related Digital Credential. Other confirmation methods can be utilized when the referenced Digital Credential supports them, in accordance with the relevant standards. | {{RFC7800}} Section 3.1, {{RFC8747}} Section 3.1 |

Reviewer 2: here, we must use the term "confirmation methods" to not use only cryptographic key bindings

Author: Resolved with the commit https://github.com/peppelinux/draft-demarco-oauth-status-attestations/pull/37/commits/ebd064bad946d0a1315417307c366642a6f625f7
